### PR TITLE
Remove PackedIntArray

### LINKED ISSUE #21676: Remove PackedIntArray from std
As the original author of the packed_int_array module, it is my opinion that it should be removed from std. My reasoning is as follows:

1. I originally wrote it to prove that we didn't need packed int arrays as a language feature during a time when Zig's semantics were quite a bit different
2. (read|write)PackedInt* are a smaller, less cumbersome implementation of much of the solution with the rest being trivially implementable from them
3. Packed structs with `@Vector`s are also effectively packed int arrays
4. There are several open issues regarding the module (#19634, #19627, #7648)
5. Of the uses of the module that I have seen, none cannot be easily replaced with the above alternatives

### PR DESCRIPTION
**Pull request by der-teufel-programming**
Fixes #21676

### CODE REVIEW

**Comment by alexrp** [comment]
Should we send it to the [orphanage](https://github.com/ziglang/std-lib-orphanage)?

**Comment by nektro** [comment]
given that part of the reasoning for proposing its removal were design flaws i dont think that's necessary. anyone who goes looking for it will still be able to find it in git

File: lib/compiler/aro/aro/Preprocessor.zig [hunk 1]
```diff
@@ -430,12 +431,24 @@ fn preprocessExtra(pp: *Preprocessor, source: Source) MacroError!TokenWithExpans
...
 
                         if (try pp.expr(&tokenizer)) {
-                            if_kind.set(if_level, until_endif);
+                            std.mem.writePackedInt(
+                                u2,
+                                &if_kind,
+                                if_level * 2,
+                                until_endif,
+                                native_endian,
+                            );
```
Comment (tgschultz) [inline lib/compiler/aro/aro/Preprocessor.zig:440]: It's a small thing, but you could use `writePackedIntNative` here to save a line in each instance.

File: lib/compiler/aro/aro/Preprocessor.zig [hunk 2]
```diff
@@ -472,23 +497,47 @@ fn preprocessExtra(pp: *Preprocessor, source: Source) MacroError!TokenWithExpans
...
+                                &if_kind,
+                                if_level * 2,
+                                until_else,
+                                native_endian,
+                            );
                         } else if (if_level == 1) {
                             guard_name = null;
                         }
-                        switch (if_kind.get(if_level)) {
+                        switch (std.mem.readPackedInt(u2, &if_kind, if_level * 2, native_endian)) {
```
Comment (tgschultz) [inline lib/compiler/aro/aro/Preprocessor.zig:532]: ..and `readPackedIntNative` in these instances

**Comment by tgschultz** [comment]
> given that part of the reasoning for proposing its removal were design flaws i don't think that's necessary. anyone who goes looking for it will still be able to find it in git

For what it is worth I agree. Anyone looking to replicate the more complicated functionality is better off implementing it from scratch using read/writePackedInt functions. I was actually going to do that, but I couldn't find any instances of anyone actually using the more complicated functionality anyway.

File: lib/compiler/aro/aro/Preprocessor.zig [hunk 3]
```diff
@@ -430,12 +430,12 @@ fn preprocessExtra(pp: *Preprocessor, source: Source) MacroError!TokenWithExpans
                         if_level = sum;
 
                         if (try pp.expr(&tokenizer)) {
-                            if_kind.set(if_level, until_endif);
+                            std.mem.writePackedIntNative(u2, &if_kind, if_level * 2, until_endif);
```
Comment (ehaas) [inline lib/compiler/aro/aro/Preprocessor.zig:433]: `if_level * 2` will overflow if it's greater than 127. I have a PR up for aro https://github.com/Vexu/arocc/pull/787 that you can grab once it's approved and passing CI (it's waiting for `@splat` for arrays to be available in pre-built zig).

**Comment by andrewrk** [comment]
Thanks!

**Comment by Atomk** [comment]
The fixed issue (https://github.com/ziglang/zig/issues/21676) links to other related issues that can probably be closed:
- https://github.com/ziglang/zig/issues/7648
- https://github.com/ziglang/zig/issues/19627

**Pull request merged**